crates/testing/tests/consensus_task.rs

```diff
@@ -120,8 +220,115 @@ async fn test_consensus_vote() {
+)]
+#[cfg_attr(async_executor_impl = "async-std", async_std::test)]
+async fn test_consensus_vote_with_permuted_dac() {
+    // Both of these tests verify that a vote is indeed sent no matter when it receives a DACRecv
+    // event. In particular, we want to verify that receiving events in an unexpected (but still
+    // valid) order allows the system to proceed as it normally would.
+    // Test the vote send where a `DACRecv` event comes first.
+    test_vote_with_specific_order(true /* dac_first */).await;
+
+    // Test the vote send where a `DACRecv` event comes second.
+    test_vote_with_specific_order(false /* dac_first */).await;
+}
+
+async fn test_vote_with_specific_order_and_panic_vote(dac_first: bool) {
+    use hotshot::tasks::{inject_consensus_polls, task_state::CreateTaskState};
+    use hotshot_task_impls::{consensus::ConsensusTaskState, events::HotShotEvent::*};
+    use hotshot_testing::{
+        predicates::{exact, is_at_view_number},
+        script::{run_test_script, TestScriptStage},
+        task_helpers::build_system_handle,
+        view_generator::TestViewGenerator,
+    };
+
+    async_compatibility_layer::logging::setup_logging();
+    async_compatibility_layer::logging::setup_backtrace();
+
+    let handle = build_system_handle(2).await.0;
+    let quorum_membership = handle.hotshot.memberships.quorum_membership.clone();
+
+    let mut generator = TestViewGenerator::generate(quorum_membership.clone());
+
+    let mut proposals = Vec::new();
+    let mut leaders = Vec::new();
+    let mut votes = Vec::new();
+    let mut dacs = Vec::new();
+    let mut vids = Vec::new();
+    for view in (&mut generator).take(2) {
+        proposals.push(view.quorum_proposal.clone());
+        leaders.push(view.leader_public_key);
+        votes.push(view.create_vote(&handle));
+        dacs.push(view.da_certificate.clone());
+        vids.push(view.vid_proposal.clone());
+    }
+
+    // Get out of the genesis view first
+    let view_1 = TestScriptStage {
+        inputs: vec![QuorumProposalRecv(proposals[0].clone(), leaders[0])],
+        outputs: vec![
+            exact(ViewChange(ViewNumber::new(1))),
+            exact(QuorumProposalValidated(proposals[0].data.clone())),
+            exact(QuorumVoteSend(votes[0].clone())),
+        ],
+        asserts: vec![is_at_view_number(1)],
+    };
+
+    let mut view_2_inputs = vec![
+        // We need a VID share for view 2 otherwise we cannot vote at view 2 (as node 2).
+        VidDisperseRecv(vids[1].0.clone(), vids[1].1),
+        DACRecv(dacs[1].clone()),
+        QuorumProposalRecv(proposals[1].clone(), leaders[1]),
+    ];
+    if !dac_first {
+        view_2_inputs.swap(1, 2);
+    }
+
+    // Now, for view 2, receive a DAC before the quorum proposal.
+    let view_2 = TestScriptStage {
+        inputs: view_2_inputs,
+        outputs: vec![
+            exact(ViewChange(ViewNumber::new(2))),
+            exact(QuorumProposalValidated(proposals[1].data.clone())),
```

Comment:
Shouldn't we expect a VoteSend?

Reply:
So this was the alternative case mentioned in the [ticket](https://github.com/EspressoSystems/HotShot/issues/2306)
>We should also verify that if we remove the vote path from one of propsal recv or dac recv event handlers that one test fails.

So this test panics because the VoteSend is missing. Let me know if I misunderstood what you meant here.

Reply:
I'm gonna remove the copy paste here and add a param instead to make this more clear.

Reply:
Done, I'll leave this unresolved to make sure that our understanding of this is the same.